neqo-http3/src/client_events.rs

```diff
@@ -268,6 +268,7 @@ impl Http3ClientEvents {
 
     /// Add a new `AuthenticationNeeded` event
     pub(crate) fn authentication_needed(&self) {
+        self.remove(|evt| matches!(evt, Http3ClientEvent::AuthenticationNeeded));
```

Comment:
Running the `http3` QUIC interop test, there is a race condition. Two `AuthenticationNeeded` events are enqueued. 

The `neqo-client` handler handles both events. 

https://github.com/mozilla/neqo/blob/9489511f7c82786f55bc9c713cddbff825507ed7/neqo-client/src/main.rs#L727-L729

On the first call to `client.authenticated` `*self.auth_required = false;` is set.

https://github.com/mozilla/neqo/blob/9489511f7c82786f55bc9c713cddbff825507ed7/neqo-crypto/src/agent.rs#L634-L643

On the second call to `client.authenticated` the assert panics `assert!(self.state.authentication_needed());`.

---

The above makes sure there is only ever a single `AuthenticationNeeded` event in the queue by removing any existing ones. Is this a valid patch? Does this change in ordering violate any invariants, i.e. does anything depend on the original order?

As an alternative, one could only call `client.authenticated` on not-yet-authenticated connections.

Alternative suggestions?

Reply:
We probably need to avoid generating that event multiple times.  I see that we generate that in response to being in the `AuthenticationPending` state on the TLS connection.  The idea must have been that the state would not be asserted multiple times (that is, we would block until someone called `authenticated()` without calling `handshake()` again).  Either that or we would de-duplicate the event successfully.

That's clearly not the case, so we're generating multiple events.  I don't see an obvious fix for this other than just tracking the state.  Try this:

```diff
diff --git a/neqo-transport/src/connection/mod.rs b/neqo-transport/src/connection/mod.rs
index 2de38841..1141765f 100644
--- a/neqo-transport/src/connection/mod.rs
+++ b/neqo-transport/src/connection/mod.rs
@@ -2578,24 +2578,30 @@ impl Connection {

     fn handshake(
         &mut self,
         now: Instant,
         packet_version: Version,
         space: PacketNumberSpace,
         data: Option<&[u8]>,
     ) -> Res<()> {
         qtrace!([self], "Handshake space={} data={:0x?}", space, data);

+        let was_authentication_pending =
+            self.crypto.tls.state() == HandshakeState::AuthenticationPending;
         let try_update = data.is_some();
         match self.crypto.handshake(now, space, data)? {
             HandshakeState::Authenticated(_) | HandshakeState::InProgress => (),
-            HandshakeState::AuthenticationPending => self.events.authentication_needed(),
+            HandshakeState::AuthenticationPending => {
+                if !was_authentication_pending {
+                    self.events.authentication_needed()
+                }
+            }
             HandshakeState::EchFallbackAuthenticationPending(public_name) => self
                 .events
                 .ech_fallback_authentication_needed(public_name.clone()),
             HandshakeState::Complete(_) => {
                 if !self.state.connected() {
                     self.set_connected(now)?;
                 }
             }
             _ => {
                 unreachable!("Crypto state should not be new or failed after successful handshake")
```